DS-2895 authorization check for item not owned by collection

abollini opened the pull request:
I don't like the cyclic dependency between ItemService, WorkspaceItem and WorkflowItem but I don't see a different way to fix the issue. The ItemService should be not responsible to implement authorization logic as in the canEdit method.
We could move the canEdit method in the authorization service but this will lead anyway to a cyclic reference (ItemService -> AuthorizationService -> ItemService / WorkspaceItemService/ WorkflowItemService).

Anyway, the cyclic dependency is not an issue, it is managed properly by Spring it is just a not nice thing that show a bad SoC

abollini committed:
DS-2895 authorization check for item not owned by collection

abollini committed:
DS-2895 wrap the updateLastModified call inside an ignore authorization block / move the call after that the action has been performed

abollini commented:
I have fixed the issue found by the WorkspaceItemTest (creation of new submission) related to an authorization exception throw by the updateLastModified call during workspaceitem creation.
To alter the resourcepolicy it was not actually necessary to have WRITE permission, it is possible only trough API and the security need to be enforced by the UI for example using a functional logic (if you can create an object you can decide the policy for such object, etc.).
Please note that I have also moved the updateLastModified call after the code that perform the resource policy change to be more consistent (before addPolicy and delete* work in a different order).

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java:
```diff
@@ -881,12 +887,28 @@ public boolean canEdit(Context context, Item item) throws SQLException {
         if (item.getOwningCollection() == null)
         {
-            return true;
+        	if (!isAnInprogressSubmission(context, item)) {
```
Why not just 'return !isAnInprogressSubmission(context, item);'?

abollini replied:
probably we should improve this logic in future...

mwoodiupui commented on dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java:
```diff
@@ -78,6 +79,11 @@
     @Autowired(required=true)
     protected HarvestedItemService harvestedItemService;
+    @Autowired(required=true)
+    protected WorkspaceItemService workspaceItemService;
```
I don't see an import for this.

abollini replied:
we have imported the full .service package, see line 21

tdonohue commented:
+1 via code review.

abollini committed:
DS-28DS-2895 add unit test to expose the wrong behaviour

slightly renamed the method isAnInprogressSubmission in isInProgressSubmission to make it uniform with DSpace 5 and added to the service interface to allow testing and easy reuse

abollini commented:
ready to merge

mwoodiupui commented:
+1 by inspection

tdonohue commented:
Unit tests provide great coverage to prove out this works right.  Approved in today's DevMtg. Merging.
